tensorboard/data/experimental/experiment_from_dev_test.py

```diff
@@ -0,0 +1,250 @@
+            self.assertEqual(request.experiment_id, "789")
+            self.assertEqual(kwargs["metadata"], grpc_util.version_metadata())
+            for run in ("train", "test"):
+                for tag in ("accuracy", "loss"):
+                    response = export_service_pb2.StreamExperimentDataResponse()
+                    response.run_name = run
+                    response.tag_name = tag
+                    display_name = "%s:%s" % (request.experiment_id, tag)
+                    response.tag_metadata.CopyFrom(
+                        test_util.scalar_metadata(display_name)
+                    )
+                    for step in range(10):
+                        response.points.steps.append(step)
+                        if tag == "loss":
+                            if run == "train":
+                                value = 1.0 / (step + 1)
+                                seconds = step
+                            else:
+                                value = -1.0 / (step + 1)
+                                seconds = 600 + step
+                        else:  # "accuracy"
+                            if run == "train":
+                                value = 1.0 / (10 - step)
+                                seconds = step * 2
+                            else:
+                                value = -1.0 / (10 - step)
+                                seconds = 600 + step * 2
+                        response.points.values.append(value)
+                        response.points.wall_times.add(seconds=seconds, nanos=0)
+                    yield response
+
+        mock_api_client.StreamExperimentData = mock.Mock(
+            wraps=stream_experiment_data
+        )
+
+        with mock.patch.object(
+            experiment_from_dev,
+            "get_api_client",
+            lambda api_endpoint: mock_api_client,
+        ):
+            experiment = experiment_from_dev.ExperimentFromDev("789")
+            dataframe = experiment.get_scalars()
+
+        # Check index.
+        train_index = dataframe.loc["train"].index
+        self.assertEqual(train_index.name, "step")
+        self.assertAllEqual(train_index, np.arange(0, 10))
+        test_index = dataframe.loc["test"].index
+        self.assertEqual(test_index.name, "step")
+        self.assertAllEqual(test_index, np.arange(0, 10))
+
+        # Check values.
+        train_losses = list(dataframe.loc["train", ("value", "loss")])
+        self.assertAllClose(train_losses, 1.0 / (np.arange(0, 10) + 1.0))
+        train_accuracies = list(dataframe.loc["train", ("value", "accuracy")])
+        self.assertAllClose(train_accuracies, 1.0 / (10.0 - np.arange(0, 10)))
+        test_losses = list(dataframe.loc["test", ("value", "loss")])
+        self.assertAllClose(test_losses, -1.0 / (np.arange(0, 10) + 1.0))
+        test_accuracies = list(dataframe.loc["test", ("value", "accuracy")])
+        self.assertAllClose(test_accuracies, -1.0 / (10.0 - np.arange(0, 10)))
```

Comment:
Would it perhaps be easier to read both the test code and failure
messages if we just construct the expected dataframe and assert that
it’s equal/close to the actual dataframe? Likewise below.

Reply:
`self.assertEqual()` doesn't work for two DataFrames. The `assertAllEqual()` and `assertAllClose()` from the `tf.test.TestCase` we're using doesn't work very well either. They succeed for equal/close DataFrames. But the big problem is that they don't check content of the indices. E.g., if you have two DataFrames with the same values and wall_times, but different run names, they won't catch the different. They'll just pass happily. Also, for failure conditions, they don't actually print out useful message about which column or which index differs between the two.

DataFrame has the `equals()` method. But it's not designed for unit testing. When two DataFrames are unequal, code like `self.assertEqual(dataframe.equals(expected))` won't tell you where the difference is.

The code here is a little long, but when failures occur, the line numbers gives you a clear idea which part of the DataFrame is the problem.

Reply:
`pandas.testing.assert_frame_equal()` works well. Done revising the tests to use it. This shortens the test code quite a lot. Thanks.